src/java.base/share/classes/java/io/ByteArrayOutputStream.java

```diff
@@ -159,8 +159,12 @@ public void writeBytes(byte[] b) {
      * @throws  NullPointerException if {@code out} is {@code null}.
      * @throws  IOException if an I/O error occurs.
      */
-    public synchronized void writeTo(OutputStream out) throws IOException {
-        out.write(buf, 0, count);
+    public void writeTo(OutputStream out) throws IOException {
+        if (Thread.currentThread().isVirtual()) {
+            out.write(toByteArray());
```

Comment:
Would it be better to avoid calling a public method `toByteArray` encase subclass is overriding it?

Reply:
A good question. The buf/count fields are protected so the subclass has direct access to the bytes. So while it could Arrays.copy the bytes, it doesn't help with a buggy subclass that is changing bytes while synchronization.

Reply:
I was thinking more of a subclass that counted invocations to public methods or metering which would cause subclass to double the counts when calling via virtual thread.

Reply:
So do we think it better not to invoke `toByteArray` here?

Reply:
Using a subclass to count the number of invocations of toByteArray seems a bit strange but in general it is more robust to not rely on a method that may be overridden by a subclass. So I think the suggestion is good.

Reply:
Currently we have
```
    public void writeTo(OutputStream out) throws IOException {
        if (Thread.currentThread().isVirtual()) {
            out.write(toByteArray());
        } else synchronized (this) {
            out.write(buf, 0, count);
        }
    }
```
where `toByteArray()` is `synchronized`, but here I would think that we'd want to replace it with simply `Arrays.copyOf(buf, count)` without the `synchronized`, no?

Reply:
@bplb My interpretation was that we didn't want to hold the monitor *during* out.write().

Reply:
Please see 8076291fb3d097ef67d0b59b9be3c8b762aad7cf.